docs/proposals/config-management-plugin-v2.md

```diff
@@ -0,0 +1,290 @@
+
+#### Improve Installation Experience
+The current Config Management plugin installation experience requires two changes:
+- An entry in configManagementPlugins in the Argo CD configmap (i.e.  argocd-cm)
+- Either an init container with a volume mount that adds a new binary into Argo CD repo server pod, or a rebuild of the argocd image, which contains the necessary tooling
+
+The problem with this approach is that the process is error-prone, manual, and requires learning from each and every Argo CD administrator. 
+
+The goal is to make additional tools easily accessible for installation to Argo CD operators.
+
+#### Provide Discovery (Auto-selection of Tool)
+For Argo CD’s natively supported config management plugins (Helm, Kustomize, Ksonnet, Jsonnet), Argo CD auto-detects 
+and selects the appropriate tool given only the path in the Git repository. 
+This selection is based on the recognition of well-known files in the directory (e.g. Chart.yaml, kustomization.yaml, etc...). 
+
+Currently, unlike natively supported tools, when a plugin is used, a user needs to explicitly specify the plugin 
+that should be used to render the manifests. As part of the improvements to config management plugins, 
+
+We want to provide the same ability to auto-select the plugin based on recognized files in the path of the git repository.
+
+### Non-Goals
+
+#### Parameters support in UI/CLI
+Currently, configuration management plugins allow specifying only a list of environment variables via UI/CLI. 
+
+We want to extend its functionality to provide a similar experience as for existing natively supported tools 
+to additional config management tools. 
+
+## Proposal
+
+We have drafted the solution to the problem statement as **running configuration management plugin tools as sidecar in the argocd-repo-server**. 
+
+All it means that Argo CD Config Management Plugin 2.0 will be,
+- A user-supplied container image with all the necessary tooling installed in it. 
+- It will run as a sidecar in the repo server deployment and will have shared access to the git repositories.
+- It will contain a CMP YAML specification file describing how to render manifests.
+- Its entrypoint will be a lightweight CMP API server that receives requests by the main repo-server to render manifests, 
+based on the CMP specification file.
+
+This mechanism will provide the following benefits over the existing solution,
+- Plugin owners control their execution environment, packaging whatever dependent binaries required.
+- An  Argo CD user who wants to use additional config management tools does not have to go through the hassle of building 
+a customized argocd-repo-server in order to install required dependencies. 
+- The plugin image will be running in a container separate from the main repo-server.
+
+### Use cases
+
+* UC1: As an Argo CD user, I would like to use first-class support provided for additional tools to generate and manage deployable kubernetes manifests
```

Comment:
If we want to support overriding one of the built-in tools via the plugin mechanism, maybe we should add an explicit use-case for that.

Reply:
Currently, we are not trying to replace support for built-in tools. This proposal aims at enhancing support for additional tools.

Reply:
We had talked in one of the contributors meeting about allowing a plugin to override one of the built-in tools, so that this new plugin mechanism could also be used to allow a different version of an existing tool to be used.  Otherwise, users would need to use the existing plugin mechanism to do that, which is more limited. @alexmt @jessesuen did we since decide not to support this?